envoy/formatter/substitution_formatter_base.h

```diff
@@ -105,35 +135,55 @@ template <class FormatterContext> class CommandParserFactoryBase : public Config
                                Server::Configuration::GenericFactoryContext& context) PURE;
 
   std::string category() const override {
-    return fmt::format("envoy.{}.formatters", FormatterContext::category());
+    static constexpr absl::string_view HttpContextCategory = "http";
+    if constexpr (formatterContextCategory<FormatterContext>() == HttpContextCategory) {
+      return "envoy.formatter"; // Backward compatibility for HTTP.
+    } else {
+      return fmt::format("envoy.formatters.{}", formatterContextCategory<FormatterContext>());
+    }
   }
 };
 
-template <class FormatterContext> class BuiltInCommandParsersBase {
+template <class FormatterContext>
+class BuiltInCommandParserFactoryBase : public Config::UntypedFactory {
 public:
-  static void addCommandParser(CommandParserBasePtr<FormatterContext> parser) {
-    mutableCommandParsers().push_back(std::move(parser));
-  }
-
-  static const CommandParsersBase<FormatterContext>& commandParsers() {
-    return mutableCommandParsers();
+  std::string category() const override {
+    return fmt::format("envoy.built_in_formatters.{}",
+                       formatterContextCategory<FormatterContext>());
   }
 
-private:
-  static CommandParsersBase<FormatterContext>& mutableCommandParsers() {
-    MUTABLE_CONSTRUCT_ON_FIRST_USE(CommandParsersBase<FormatterContext>);
-  }
+  /**
+   * Creates a particular CommandParser implementation.
+   */
+  virtual CommandParserBasePtr<FormatterContext> createCommandParser() const PURE;
 };
 
-template <class FormatterContext> struct BuiltInCommandPaserRegister {
-  BuiltInCommandPaserRegister(CommandParserBasePtr<FormatterContext> parser) {
-    BuiltInCommandParsersBase<FormatterContext>::addCommandParser(std::move(parser));
+using BuiltInStreamInfoCommandParserFactory = BuiltInCommandParserFactoryBase<void>;
+
+/**
+ * Helper class to get all built-in command parsers for a given formatter context.
+ */
+template <class FormatterContext> class BuiltInCommandParserFactoryHelper {
+public:
+  using Factory = BuiltInCommandParserFactoryBase<FormatterContext>;
+  using Parsers = std::vector<CommandParserBasePtr<FormatterContext>>;
+
+  /**
+   * Get all built-in command parsers for a given formatter context.
+   * @return Parsers all built-in command parsers for a given formatter context.
+   */
+  static const Parsers& commandParsers() {
+    CONSTRUCT_ON_FIRST_USE(Parsers, []() {
+      Parsers parsers;
+      for (auto& f : Registry::FactoryRegistry<Factory>::factories()) {
+        auto parser = f.second->createCommandParser();
+        RELEASE_ASSERT(parser != nullptr, "Null built-in command parser");
```

Comment:
Can this be an envoy bug and no push back?

Reply:
SGTM.